src/core.c/Channel.pm6

```diff
@@ -103,11 +103,11 @@ my class Channel does Awaitable {
             Nil
         } else {
             if nqp::istype(msg, CHANNEL_CLOSE) {
-                $!closed_promise_vow.keep(Nil);
+                try $!closed_promise_vow.keep(Nil);
```

Comment:
`try` is actually pretty expensive thing. I think it'd be more efficient to have something like:

```
$!closed_promise_vow.keep(Nil) unless $!closed_promise_vow
```

Reply:
That would be a data race, wouldn't it?

Normally the `try` will be encountered once for each `Channel` instance. Is there a common use case where many short-lived channels are created and a slow `try` would have a noticeable impact?

Anyway, I see only a few ways to avoid the `try`:

1.  ensure that `!keep` encounters a `CHANNEL_CLOSE`/`CHANNEL_FAIL` only once -- I don't know how to accomplish that without some kind of mutex around the `$!closed` variable
2. add some kind of `:ignore-resolved` option to the `keep`/`break` methods of the vow -- I don't like that approach much but maybe there is a precedent in core?
3. have the `keep`/`break` methods of the vow `fail` instead of throwing -- that seems feasible and would still allow detection of attempts to resolve a promise a second time

Reply:
You're right, since `Channel` itself isn't locking my suggestion is really unsafe. Besides, I can think of a couple of scenarios where this code path would be used repeatedly. But none of them is from a book about good programming practices. :)

As to use of `fail`, I was considering it. (Though only for cases when `Promise` was about to change into the same state it is in already. I.e. kept -> kept would be a failure; but kept -> broken is an immediate throw.) What I don't like is that `Failure` is an action at distance. It's not really good for cases when method return type is usually ignored.

So, in fact, everything is good here. Thank you!